sql: ensure that AS OF SYSTEM TIME is handled consistently

knz opened the pull request:
Fixes #20182.

Prior to this patch, two inconsistencies were present in the code:

- a user-visible inconsistency: `AS OF SYSTEM TIME` (henceforth
  referred to as "AOST") was accepted anywhere in a query as soon as
  it was present at the top level SELECT statement, including with
  conflicting timestamps. Accepting AOST in multiple places doesn't
  feel undesirable, but allowing conflicting timestamps to be
  specified is unsound with the current underlying mechanism (a shared
  timestamp for the entire transaction).

- an internal inconsistency, visible to CockroachDB developers: the
  presence of an AOST clause during planning was erroneously conflated
  with the flag that disables caching of table descriptors
  (`planner.avoidCachedDescriptors`). This is erroneous because,
  although AOST *implies* `avoidCachedDescriptors == true` (we can't
  use the cache while time travelling), the converse is not true: when
  processing view descriptors (and perhaps, in the future, for other
  reasons), we also disable descriptor caching although AOST is not
  involved.

This patch rectifies this situation as follows:

- a new planner flag `asOfSystemTime` is introduced to indicate
  that an AOST clause was properly recognized at the top level.

- the logic that allows or refuses AOST clauses in FROM clauses in
  arbitrarily nested SELECT clauses (including, potentially, those
  expanded from views), is modified to use this new flag.

- the time stamps of AOST clauses, if multiple are specified, are
  checked to be equal to the one set at the top level. This
  restriction might be lifted in the future if we ever support
  different AOST clauses per data source.

In addition, the error message when an AOST clause is not given in the
proper syntactic position is improved to hint where it should be
placed instead.

----

Release note (sql, bug fix): it is not possible any more to indicate
conflicting `AS OF SYSTEM TIME` clauses in different part of a query.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20267)
<!-- Reviewable:end -->

knz commented:
Might cherry pick this on 1.1. what do you think?

knz committed:
sql: ensure that AS OF SYSTEM TIME is handled consistently

Prior to this patch, two inconsistencies were present in the code:

- a user-visible inconsistency: `AS OF SYSTEM TIME` (henceforth
  referred to as "AOST") was accepted anywhere in a query as soon as
  it was present at the top level SELECT statement, including with
  conflicting timestamps. Accepting AOST in multiple places doesn't
  feel undesirable, but allowing conflicting timestamps to be
  specified is unsound with the current underlying mechanism (a shared
  timestamp for the entire transaction).

- an internal inconsistency, visible to CockroachDB developers: the
  presence of an AOST clause during planning was erroneously conflated
  with the flag that disables caching of table descriptors
  (`planner.avoidCachedDescriptors`). This is erroneous because,
  although AOST *implies* `avoidCachedDescriptors == true` (we can't
  use the cache while time travelling), the converse is not true: when
  processing view descriptors (and perhaps, in the future, for other
  reasons), we also disable descriptor caching although AOST is not
  involved.

This patch rectifies this situation as follows:

- a new planner flag `asOfSystemTime` is introduced to indicate
  that an AOST clause was properly recognized at the top level.

- the logic that allows or refuses AOST clauses in FROM clauses in
  arbitrarily nested SELECT clauses (including, potentially, those
  expanded from views), is modified to use this new flag.

- the timestamps of AOST clauses, if multiple are specified, are
  checked to be equal to the one set at the top level. This
  restriction might be lifted in the future if we ever support
  different AOST clauses per data source.

In addition, the error message when an AOST clause is not given in the
proper syntactic position is improved to hint where it should be
placed instead.

----

Release note (sql, bug fix): it is not possible any more to indicate
conflicting `AS OF SYSTEM TIME` clauses in different part of a query.

mjibson commented:
1.1 cherrypick SGTM.

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 539 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20267#-KzzDefLIFu8ya2W9h5u:-KzzDefLIFu8ya2W9h5v:b-vp3ir5) ([raw file](https://github.com/cockroachdb/cockroach/blob/674e4b9ef54336ddd26e28179ecd3812308b371b/pkg/sql/executor.go#L539)):*
> ```Go
> 		// we want the descriptors at the timestamp given, not the latest
> 		// known to the cache.
> 		planner.avoidCachedDescriptors = true
> ```

Isn't `avoidCachedDescriptors` FUD by now? @vivekmenezes is there anything left to be done for it to go away?

---

*[pkg/sql/planner.go, line 62 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20267#-KzzE-NiwwnawPoHUaJt:-KzzE-NiwwnawPoHUaJu:b-11umgr) ([raw file](https://github.com/cockroachdb/cockroach/blob/674e4b9ef54336ddd26e28179ecd3812308b371b/pkg/sql/planner.go#L62)):*
> ```Go
> 	// txn.mu.Proto.OrigTimestamp). If set, avoidCachedDescriptors below
> 	// must also be set.
> 	// TODO(anyone): we may want to support table readers at arbitrary
> ```

future nit: As per the google styleguide, the name appearing in the TODO is the name of a person who can answer questions about it. TODO(anyone) doesn't make sense.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20267#-:-KzzF1Nd0U6Urp2BJMYt:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 539 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20267#-KzzDefLIFu8ya2W9h5u:-KzzJUMW908RemFVOfIy:bhdt7xy) ([raw file](https://github.com/cockroachdb/cockroach/blob/674e4b9ef54336ddd26e28179ecd3812308b371b/pkg/sql/executor.go#L539)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Isn't `avoidCachedDescriptors` FUD by now? @vivekmenezes is there anything left to be done for it to go away?
</blockquote></details>

Wait what? It changes all the descriptor lookup paths (used throughout planning) to use either the descriptor cache or not. It would be blatantly incorrect to use the cache for things like AOST, view descriptors etc.

---

*[pkg/sql/planner.go, line 62 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20267#-KzzE-NiwwnawPoHUaJt:-KzzJczhzuc_HHhiFcqC:bars3ol) ([raw file](https://github.com/cockroachdb/cockroach/blob/674e4b9ef54336ddd26e28179ecd3812308b371b/pkg/sql/planner.go#L62)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

future nit: As per the google styleguide, the name appearing in the TODO is the name of a person who can answer questions about it. TODO(anyone) doesn't make sense.
</blockquote></details>

Well noted, thanks for the feedback.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20267)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 539 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20267#-KzzDefLIFu8ya2W9h5u:-KzzKMvm6mbC6wVqh0FZ:b-zfmwbs) ([raw file](https://github.com/cockroachdb/cockroach/blob/674e4b9ef54336ddd26e28179ecd3812308b371b/pkg/sql/executor.go#L539)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Wait what? It changes all the descriptor lookup paths (used throughout planning) to use either the descriptor cache or not. It would be blatantly incorrect to use the cache for things like AOST, view descriptors etc.
</blockquote></details>

All queries have a timestamp and the caches are supposed to be aware of that; AOST queries are not supposed to be special in that regard. At least that's what the idea is...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20267)*
<!-- Sent from Reviewable.io -->

knz commented:
Oh I see. Perhaps we could just try nuking it and see what happens. 
I would however recommend a casual investigation at the commit that introduced the avoidCachedDescriptor hack for view handling. I recall it wasn't there before and I needed to add it to fix a bug, but I don't remember which bug.

vivekmenezes commented:
we can certainly use the table descriptor cache for storing descriptors used by such queries. It's just never been a priority because these queries have tended to not be OLTP ones.

vivekmenezes commented:
And yeah it will make the code a lot cleaner by eliminating the need to care about this. So that's a good reason to want to do it.
